src/uid2ApiClient.ts

```diff
@@ -1,122 +1,189 @@
-    if (isUnvalidatedRefreshResponse(response)) {
-        return response.status === "optout" || response.status === "expired_token" ||
-            (response.status === "success" && 'body' in response && isValidIdentity(response.body));
+type RefreshApiResponse =
+  | {
+      status: ResponseStatusRequiringBody;
+      body: Uid2Identity;
     }
-    return false;
+  | {
+      status: ResponseStatusWithoutBody;
+    };
+export type ResponseStatus =
+  | ResponseStatusRequiringBody
+  | ResponseStatusWithoutBody;
+type ResponseStatusRequiringBody = "success";
+type ResponseStatusWithoutBody = "optout" | "expired_token";
+type UnvalidatedRefreshResponse = RefreshApiResponse | { status: unknown };
+function isValidRefreshResponse(
+  response: unknown | UnvalidatedRefreshResponse
+): response is RefreshApiResponse {
+  if (isUnvalidatedRefreshResponse(response)) {
+    return (
+      response.status === "optout" ||
+      response.status === "expired_token" ||
+      (response.status === "success" &&
+        "body" in response &&
+        isValidIdentity(response.body))
+    );
+  }
+  return false;
 }
-function isUnvalidatedRefreshResponse(response: unknown): response is UnvalidatedRefreshResponse {
-    return typeof(response) === 'object' && response !== null && 'status' in response;
+function isUnvalidatedRefreshResponse(
+  response: unknown
+): response is UnvalidatedRefreshResponse {
+  return (
+    typeof response === "object" && response !== null && "status" in response
+  );
 }
 export type Uid2ApiClientOptions = {
-    baseUrl?: string;
+  baseUrl?: string;
 };
 
 export class Uid2ApiClient {
-    private _baseUrl: string;
-    private _clientVersion: string;
-    private _requestsInFlight: XMLHttpRequest[] = [];
-    constructor(opts: Uid2ApiClientOptions) {
-        this._baseUrl = opts.baseUrl ?? "https://prod.uidapi.com";
-        this._clientVersion = 'uid2-sdk-' + UID2.VERSION;
-    }
+  private _baseUrl: string;
+  private _clientVersion: string;
+  private _requestsInFlight: XMLHttpRequest[] = [];
+  public _refreshingPromise:
+    | {
+        resolve: (value: string) => void;
+        reject: (reason: Error | string) => void;
+      }
+    | undefined;
+
+  constructor(opts: Uid2ApiClientOptions, uid2Sdk?: UID2) {
+    this._baseUrl = opts.baseUrl ?? "https://prod.uidapi.com";
+    this._clientVersion = "uid2-sdk-" + UID2.VERSION;
+    if (uid2Sdk) uid2Sdk.callbacks.push(this._handleEvent.bind(this));
+  }
 
-    private createArrayBuffer(text: string) {
-        const arrayBuffer = new Uint8Array(text.length);
-        for (let i = 0; i < text.length; i++) {
-            arrayBuffer[i] = text.charCodeAt(i);
+  public getFreshAdvertisingToken() {
+    return new Promise<string>((resolve, reject) => {
+      this._refreshingPromise = { resolve: resolve, reject: reject };
+    });
+  }
+
+  private _handleEvent(eventType: EventType, payload: Uid2CallbackPayload) {
```

Comment:
There should be a simpler way to do this - I don't like introducing this extra responsibility here. The SDK client is really just that - a client. It doesn't know about the SDK, or integrations, and it's likely to be used in places where we aren't using that (e.g. Prebid.js).

The existing data flow for renewing a token looks roughly like this:
![image](https://user-images.githubusercontent.com/111951647/208565051-444086d6-23aa-48bc-ae5b-e53e5a981d35.png)

With this change, there's an extra loop back down to this class after it's been through the event system:
![image](https://user-images.githubusercontent.com/111951647/208565200-33ac984e-5a9a-43b9-bf9c-59423b30420f.png)

You could move all of these changes somewhere else easily - the only thing they need to know about the API client is whether there's an active request, and you can just use `.hasActiveRequests()` to check that. That way you can avoid the extra dependency here.

Reply:
I've now put those changes into uid2PromiseHandler, which is used to take care of the `getAdvertisingTokenAsync`, as we now change the behavior of `getAdvertisingTokenAsync`, I think it make sense to still let uid2PromiseHandler to hold those logic